kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CuboidStatisticsJob.scala

```diff
@@ -0,0 +1,172 @@
+ */
+object CuboidStatisticsJob {
+
+  /**
+   * @param inputDs Part of FlatTable which contains all normal dimensions
+   * @return Cuboid level statistics data
+   */
+  def statistics(inputDs: Dataset[Row], seg: SegmentInfo): Array[(String, AggInfo)] = {
+
+    val rkc = seg.allColumns.count(c => c.rowKey)
+    // maybe we should use sample operation to reduce cost later
+    val res = inputDs.rdd
+      .mapPartitions(new CuboidStatisticsJob(seg.getAllLayout.map(x => x.getId), rkc).statisticsWithinPartition)
+    val l = res.map(a => (a.key, a)).reduceByKey((a, b) => a.merge(b)).collect()
+    //    l.foreach(x => println(x._1 + " >>><<< " + x._2.cuboid.counter.getCountEstimate))
+    l
+  }
+}
+
+class CuboidStatisticsJob(ids: List[Long], rkc: Int) extends Serializable {
+  private val info = mutable.Map[String, AggInfo]()
+  private var allCuboidsBitSet: Array[Array[Integer]] = Array()
+  private val hf: HashFunction = Hashing.murmur3_128
+  private val rowHashCodesLong = new Array[Long](rkc)
+  private var idx = 0
+  private var meter1 = 0L
+  private var meter2 = 0L
+  private var startMills = 0L
+  private var endMills = 0L
+
+
+  def statisticsWithinPartition(rows: Iterator[Row]): Iterator[AggInfo] = {
+    init()
+    println("CuboidStatisticsJob-statisticsWithinPartition1-" + System.currentTimeMillis())
+    rows.foreach(update)
+    printStat()
+    println("CuboidStatisticsJob-statisticsWithinPartition2-" + System.currentTimeMillis())
+    info.valuesIterator
+  }
+
+  def init(): Unit = {
+    println("CuboidStatisticsJob-Init1-" + System.currentTimeMillis())
+    allCuboidsBitSet = getCuboidBitSet(ids, rkc)
+    ids.foreach(i => info.put(i.toString, AggInfo(i.toString)))
+    println("CuboidStatisticsJob-Init2-" + System.currentTimeMillis())
+  }
+
+  def update(r: Row): Unit = {
+    idx += 1
+    if (idx <= 5)
+      println(r)
+    updateCuboid(r)
+  }
+
+  def updateCuboid(r: Row): Unit = {
+    // generate hash for each row key column
+    startMills = System.currentTimeMillis()
+    var idx = 0
+    while (idx < rkc) {
+      val hc = hf.newHasher
+      var colValue = r.get(idx).toString
+      if (colValue == null) colValue = "0"
+      // add column ordinal to the hash value to distinguish between (a,b) and (b,a)
+      rowHashCodesLong(idx) = hc.putUnencodedChars(colValue).hash().padToLong() + idx
+      idx += 1
+    }
+    endMills = System.currentTimeMillis()
+    meter1 += (endMills - startMills)
+
+
+    startMills = System.currentTimeMillis()
+    // use the row key column hash to get a consolidated hash for each cuboid
+    val n = allCuboidsBitSet.length
+    idx = 0
+    while (idx < n) {
+      var value: Long = 0
+      var position = 0
+      while (position < allCuboidsBitSet(idx).length) {
+        value += rowHashCodesLong(allCuboidsBitSet(idx)(position))
+        position += 1
+      }
+      info(ids(idx).toString).cuboid.counter.addHashDirectly(value)
```

Comment:
If we have 10 thousand cuboid to be statistics, we will call Long.toString 10 thousand times and create 10 thousand String objects. Is it the cause of bad performance ?